[Inductor] Account for shared memory from addmm, persistent in gemm config pruning

PaulZhang12 opened the pull request:
Differential Revision: D89510915

TMA persistent template generally results in more shared memory from realizing the epilogue, due to layout conversions in shared memory for the epilogue while other loads are happening. Example for addmm persistent: https://pastebin.com/y1ezWP88

Addmm does the layout conversion in the acc dtype (fp32) for adding onto the bias afterwards while mm does it in the output dtype before calling tl.store.

The shared memory checker previously did not prune enough, which ended up providing out of memory configs potentially when the flag was on. This PR makes a few changes to the logic:
* Disables shared memory checker by default
* Instead of under-pruning, where pruning only occurred when the shared memory of the loads were greater than the shared memory available, a poor heuristic as pipelining assumptions, etc. did not always hold true due to optimizations, we now over-prune, where the theoretical maximum shared memory is taken and if it is larger than the shared memory available, pruning occurs. This results in a few good configs being pruned, but ensures that all potentially out of memory configs no longer exist
* Exhaustive autotuning does not use the shared memory checker due to potentially pruning good configs




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

njriasan reviewed (commented):
@PaulZhang12 Not all of these are blockers, but I think we need a better abstraction here. Overall the content looks good though.

njriasan commented on torch/_inductor/template_heuristics/triton.py:
```diff
@@ -653,6 +653,8 @@ def _scale_mm_configs(
     def _get_exceeding_shared_memory_checker(
         self,
+        has_sm_layout_conversion: bool = False,
+        layout_conversion_byte_size: int = 0,
```
I like conceptually separating these, but I can't help but notice all the math works out if 0 maps to False.

njriasan commented on torch/_inductor/template_heuristics/triton.py:
```diff
@@ -226,6 +226,11 @@ def __call__(
 
+# Padding factor for shared memory estimator to account for
+# small shared memory usage of kernels, e.g. barriers
+SHARED_MEM_ESTIMATION_PADDING_FACTOR = 1.02
```
Is this necessary? Assume this only matters near the limit, so we have say 200,000 bytes allocated. 2% of that is 4000, which seems high for barriers.

I'm guessing you add this so you can prove the actual shared memory is strictly less?

njriasan commented on torch/_inductor/template_heuristics/triton.py:
```diff
@@ -790,11 +795,40 @@ def _scale_mm_configs(
+        # is not free and takes additional shared memory, while next loads are prefetched
+        # For addmm, the conversion is in the acc dtype, as it is needed before the bias addition
+        # For mm, the conversion is in the output dtype, as it happens before the store
+        if has_sm_layout_conversion:
```
How do we determine/estimate this. Getting this right seems essential to the pruning assumption.

PaulZhang12 committed:
[Inductor] Account for addmm + persistent shared memory use for config pruning (#170832)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/170832

Test Plan: `buck2 run mode/opt caffe2/test/inductor:max_autotune -- -r test_addmm_shared_memory_pruning`

Reviewed By: njriasan

Differential Revision: D89510915
